### ISSUE TITLE
Failed to send_join via any server for #emacs:matrix.org

### ISSUE THREAD

**Issue by MTRNord** [matrix-org/synapse#10021]
### Description

Joining #emacs:matrix.org fails with a `Failed to send_join via any server`. Join method was the room directory via element web.

### Steps to reproduce

- Search emacs in the matrix.org room dir via ele web
- Click and join the room

<!--
Describe how what happens differs from what you expected.

If you can identify any relevant log snippets from _homeserver.log_, please include
those (please be careful to remove any personal or private data). Please surround them with
``` (three backticks, on a line on their own), so that they are formatted legibly.
-->

```
2021-05-19 16:19:31,049 - synapse.federation.federation_client - 548 - WARNING - POST-1001498 - Failed to send_join via matrix.org
Traceback (most recent call last):
  File "/opt/synapse/synapse_git/synapse/federation/federation_client.py", line 517, in _try_destination_list
    return await callback(destination)
  File "/opt/synapse/synapse_git/synapse/federation/federation_client.py", line 711, in send_request
    room_version=room_version,
  File "/opt/synapse/synapse_git/synapse/federation/federation_client.py", line 427, in _check_sigs_and_hash_and_fetch
    ).addErrback(unwrapFirstError)
  File "/opt/synapse/env/lib/python3.7/site-packages/twisted/internet/defer.py", line 761, in send
    raise result.value
twisted.internet.defer.FirstError: FirstError[#0, [Failure instance: Traceback: <class 'synapse.api.errors.SynapseError'>: 403: event id $1449445060123025NBoor:matrix.org: unable to verify signature for sender ghostcorp.net: 400: Not signed by ghostcorp.net
/opt/synapse/synapse_git/synapse/federation/federation_base.py:76:_check_sigs_and_hashes
/opt/synapse/synapse_git/synapse/federation/federation_base.py:219:_check_sigs_on_pdus
/opt/synapse/env/lib/python3.7/site-packages/twisted/internet/defer.py:348:addErrback
/opt/synapse/env/lib/python3.7/site-packages/twisted/internet/defer.py:330:addCallbacks
--- <exception caught here> ---
/opt/synapse/env/lib/python3.7/site-packages/twisted/internet/defer.py:662:_runCallbacks
/opt/synapse/synapse_git/synapse/federation/federation_base.py:216:sender_err
]]
```

Which seems to suggest that room is broken from the perspective of my HS?

This has the side effect that my voyager bot tries to join the room over and over, causes a membership event but never actually joins from my synapse perspective.


Perspective from other servers is that synapse sends membership join events over and over again. On Element android this spams the timeline and as my synapse doesnt realise this I cannot really prevent this in my code I think unless I cache this data myself.


### Version information

<!-- IMPORTANT: please answer the following questions, to help us narrow down the problem -->

<!-- Was this issue identified on matrix.org or another homeserver? -->
- **Homeserver**: matrix.nordgedanken.dev

If not matrix.org:

<!--
 What version of Synapse is running?

You can find the Synapse version with this command:

$ curl http://localhost:8008/_synapse/admin/v1/server_version

(You may need to replace `localhost:8008` if Synapse is not configured to
listen on that port.)
-->
- **Version**: 1.34.0 (b=develop,9752849e2,dirty)

- **Install method**: Git clone with pip
<!-- examples: package manager/git clone/pip  -->

- **Platform**: Debian 10
<!--
Tell us about the environment in which your homeserver is operating
distro, hardware, if it's running in a vm/container, etc.
-->

**Comment by MTRNord** [comment]
View from other Homeservers in the emacs room:

![image](https://user-images.githubusercontent.com/1374914/118833923-22168700-b8c2-11eb-843f-c3a14b9fd37c.png)

**Comment by erikjohnston** [comment]
That stack trace is really odd, we should be catching that error correctly:

https://github.com/matrix-org/synapse/blob/1350b053da45c94722cd8acf9cfd367db787259c/synapse/federation/federation_client.py#L391-L394

**Comment by MTRNord** [comment]
@erikjohnston Not if this makes any difference but I have a workaround too applied that fixes https://github.com/matrix-org/synapse/issues/9956 for me. I do multiple by 10000 in this line: https://github.com/matrix-org/synapse/blob/3ff225175462dde8376aa584e3a47c43b1f0e790/synapse/http/site.py#L106 but I dont think this makes a difference in this very issue. But I thought it is better to mention than to not mention it.

**Comment by Hodapp87** [comment]
Seeing the same thing on a fresh installation of 1.34.0.

**Comment by Necronian** [comment]
I've had the issue since early December.

**Comment by hiredman** [comment]
I see the same error when trying to join #fedora:matrix.org also referencing ghostcorp.net

**Comment by mkroman** [comment]
I'm having a similar issue when trying to join #ghidra:matrix.org on synapse v1.38.0.

```
2021-07-15 06:04:35,228 - synapse.access.http.8008 - 389 - INFO - GET-67- 10.42.1.122 - 8008 - {@mk:rwx.im} Processed request: 30.138sec/-0.000sec (0.001sec, 0.000sec) (0.001sec/0.130sec/1) 473B 200 "GET /_matrix/client/r0/sync?filter=0&timeout=30000&since=s19082_198914_0_3360_451_5_82_8486_1 HTTP/1.1" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Element/1.7.29 Chrome/89.0.4389.128 Electron/12.0.14 Safari/537.36" [0 dbevts]
2021-07-15 06:04:36,755 - synapse.http.matrixfederationclient - 278 - INFO - POST-71- {GET-O-18} [matrix.org] Completed request: 200 OK in 0.14 secs, got 69 bytes - GET matrix://matrix.org/_matrix/federation/v1/query/directory?room_alias=%23ghidra%3Amatrix.org
2021-07-15 06:04:37,283 - synapse.http.matrixfederationclient - 278 - INFO - POST-71- {GET-O-19} [matrix.org] Completed request: 200 OK in 0.25 secs, got 1056 bytes - GET matrix://matrix.org/_matrix/federation/v1/make_join/%21RistbroappQPsmzqnr%3Amatrix.org/%40mk%3Arwx.im?ver=1&ver=2&ver=3&ver=4&ver=5&ver=6&ver=org.matrix.msc2176&ver=org.matrix.msc3083&ver=7
2021-07-15 06:04:38,520 - synapse.metrics - 598 - INFO - sentinel- Collecting gc 1
2021-07-15 06:04:43,870 - synapse.http.matrixfederationclient - 278 - INFO - POST-71- {PUT-O-20} [matrix.org] Completed request: 200 OK in 6.55 secs, got 11110268 bytes - PUT matrix://matrix.org/_matrix/federation/v2/send_join/%21RistbroappQPsmzqnr%3Amatrix.org/%2416263290773VTuNy%3Arwx.im
2021-07-15 06:04:43,870 - synapse.federation.federation_client - 732 - INFO - POST-71- Processing from send_join 9706 events
2021-07-15 06:04:44,502 - synapse.federation.federation_client - 575 - WARNING - POST-71- Failed to send_join via matrix.org
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/twisted/internet/defer.py", line 1445, in _inlineCallbacks
    result = current_context.run(g.send, result)
StopIteration: 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/twisted/internet/defer.py", line 1445, in _inlineCallbacks
    result = current_context.run(g.send, result)
StopIteration: {'ed25519:auto': FetchKeyResult(verify_key=<nacl.signing.VerifyKey object at 0x7ff6039cf340>, valid_until_ts=1576767829750), 'ed25519:a_RXGa': FetchKeyResult(verify_key=<nacl.signing.VerifyKey object at 0x7ff6039da1f0>, valid_until_ts=1626404282697)}

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 545, in _try_destination_list
    return await callback(destination)
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 752, in send_request
    await concurrently_execute(
  File "/usr/local/lib/python3.8/site-packages/twisted/internet/defer.py", line 1445, in _inlineCallbacks
    result = current_context.run(g.send, result)
  File "/usr/local/lib/python3.8/site-packages/synapse/util/async_helpers.py", line 187, in _concurrently_execute_inner
    await maybe_awaitable(func(value))
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 742, in _execute
    valid_pdu = await self._check_sigs_and_hash_and_fetch_one(
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 432, in _check_sigs_and_hash_and_fetch_one
    res = await self._check_sigs_and_hash(room_version, pdu)
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_base.py", line 65, in _check_sigs_and_hash
    if not check_event_content_hash(pdu):
  File "/usr/local/lib/python3.8/site-packages/synapse/crypto/event_signing.py", line 42, in check_event_content_hash
    name, expected_hash = compute_content_hash(event.get_pdu_json(), hash_algorithm)
  File "/usr/local/lib/python3.8/site-packages/synapse/crypto/event_signing.py", line 96, in compute_content_hash
    event_json_bytes = encode_canonical_json(event_dict)
  File "/usr/local/lib/python3.8/site-packages/canonicaljson.py", line 73, in encode_canonical_json
    s = _canonical_encoder.encode(json_object)
  File "/usr/local/lib/python3.8/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/local/lib/python3.8/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/local/lib/python3.8/site-packages/canonicaljson.py", line 30, in _default
    raise TypeError(
TypeError: Object of type Decimal is not JSON serializable
2021-07-15 06:04:44,503 - synapse.http.server - 88 - INFO - POST-71- <XForwardedForRequest at 0x7ff60ad854c0 method='POST' uri='/_matrix/client/r0/join/%23ghidra%3Amatrix.org' clientproto='HTTP/1.1' site='8008'> SynapseError: 502 - Failed to send_join via any server```

**Comment by dada513** [comment]
same issue #linux:libera.chat

**Comment by benparsons** [comment]
I'm have the same issue:

```
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]: 2021-08-09 13:39:51,844 - synapse.federation.federation_client - 575 - WARNING - POST-1250672 - Failed to make_join via vector.modular.im
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]: Traceback (most recent call last):
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:   File "/usr/local/lib/python3.8/site-packages/twisted/internet/defer.py", line 1661, in _inlineCallbacks
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:     result = current_context.run(gen.send, result)
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]: StopIteration: DestinationRetryTimings(failure_ts=1628516389382, retry_last_ts=1628516389382, retry_interval=600000)
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]: During handling of the above exception, another exception occurred:
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]: Traceback (most recent call last):
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:   File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 545, in _try_destination_list
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:     return await callback(destination)
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:   File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 632, in send_request
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:     ret = await self.transport_layer.make_membership_event(
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:   File "/usr/local/lib/python3.8/site-packages/synapse/federation/transport/client.py", line 243, in make_membership_event
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:     content = await self.client.get_json(
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:   File "/usr/local/lib/python3.8/site-packages/synapse/http/matrixfederationclient.py", line 1007, in get_json
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:     response = await self._send_request_with_optional_trailing_slash(
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:   File "/usr/local/lib/python3.8/site-packages/synapse/http/matrixfederationclient.py", line 391, in _send_request_with_optional_trailing_slash
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:     response = await self._send_request(request, **send_request_args)
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:   File "/usr/local/lib/python3.8/site-packages/synapse/http/matrixfederationclient.py", line 479, in _send_request
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:     limiter = await synapse.util.retryutils.get_retry_limiter(
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:   File "/usr/local/lib/python3.8/site-packages/synapse/util/retryutils.py", line 92, in get_retry_limiter
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]:     raise NotRetryingDestination(
Aug 09 13:39:51 benp-synapse matrix-synapse[2552]: synapse.util.retryutils.NotRetryingDestination: Not retrying server vector.modular.im.
```

**Comment by MTRNord** [comment]
@benparsons looking at your trace its a different issue causing the same error? As mine is a missing a signature while yours is not retrying the server? 🤔

**Comment by TheGreatestJannet** [comment]
Is there a fix for this? I can't join `#nextcloud:matrix.org` at all.
I also get the following error:
```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/twisted/internet/defer.py", line 1661, in _inlineCallbacks
    result = current_context.run(gen.send, result)
StopIteration: {'ed25519:auto': FetchKeyResult(verify_key=<nacl.signing.VerifyKey object at 0x7f1dc06d7d90>, valid_until_ts=1576767829750), 'ed25519:a_RXGa': FetchKeyResult(verify_key=<nacl.signing.VerifyKey object at 0x7f1dc45c2100>, valid_until_ts=1636323490080)}

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 576, in _try_destination_list
    return await callback(destination)
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 825, in send_request
    await concurrently_execute(
  File "/usr/local/lib/python3.8/site-packages/twisted/internet/defer.py", line 1661, in _inlineCallbacks
    result = current_context.run(gen.send, result)
  File "/usr/local/lib/python3.8/site-packages/synapse/util/async_helpers.py", line 192, in _concurrently_execute_inner
    await maybe_awaitable(func(value))
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 815, in _execute
    valid_pdu = await self._check_sigs_and_hash_and_fetch_one(
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 458, in _check_sigs_and_hash_and_fetch_one
    res = await self._check_sigs_and_hash(room_version, pdu)
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_base.py", line 70, in _check_sigs_and_hash
    if not check_event_content_hash(pdu):
  File "/usr/local/lib/python3.8/site-packages/synapse/crypto/event_signing.py", line 42, in check_event_content_hash
    name, expected_hash = compute_content_hash(event.get_pdu_json(), hash_algorithm)
  File "/usr/local/lib/python3.8/site-packages/synapse/crypto/event_signing.py", line 96, in compute_content_hash
    event_json_bytes = encode_canonical_json(event_dict)
  File "/usr/local/lib/python3.8/site-packages/canonicaljson.py", line 83, in encode_canonical_json
    s = _canonical_encoder.encode(json_object)
  File "/usr/local/lib/python3.8/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/local/lib/python3.8/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/local/lib/python3.8/site-packages/canonicaljson.py", line 36, in _default
    raise TypeError(
TypeError: Object of type Decimal is not JSON serializable
```

**Comment by anoadragon453** [comment]
```
TypeError: Object of type Decimal is not JSON serializable
```

errors should be fixed in the next release of Synapse, v1.47.0. That, however, is a different error from what was originally posted, which was the result of event signature verification failure.

**Comment by richvdh** [comment]
This issue seems to have become a bit of a pile-on for everyone who ever saw any sort of error joining any sort of room. I was able to join `#emacs:matrix.org` with no problem, so I think the original problem has probably been solved, and I'm going to close this.

If people have problems joining rooms, please create a *new* issue rather than piling onto one that sounds vaguely related - there are a lot of reasons a join can fail and it's much easier for us to close duplicate issues than it is to split out unrelated problems.
